File: python/e2e_harness/core/env.py
"""Environment configuration loader.

Reads config.yaml with ${VAR:-default} interpolation, falling back to
environment variables when config.yaml is absent.
"""

from __future__ import annotations

import os
import re
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

import yaml


@dataclass
class ServiceConfig:
    name: str = ""
    base_url: str = ""


@dataclass
class AuthConfig:
    tenant_id: str = ""
    user_id: str = ""
    headers: dict[str, str] = field(default_factory=dict)
    extra: dict[str, str] = field(default_factory=dict)


@dataclass
class RuntimeConfig:
    http_timeout_s: int = 120
    poll_interval_ms: int = 500
    poll_timeout_s: int = 60
    parallel: int = 4


@dataclass
class Capabilities:
    known: bool = False
    data: dict[str, Any] = field(default_factory=dict)


@dataclass
class DiscoverConfigSection:
    """Optional dev-time tooling config for discover/scaffold."""

    source_root: str = ""  # dir to scan for @case/@spec (relative to config.yaml dir)
    test_root: str = ""  # where tests land, grouped <test_root>/<group>/ (relative to config.yaml dir)


@dataclass
class Env:
    service: ServiceConfig = field(default_factory=ServiceConfig)
    auth: AuthConfig = field(default_factory=AuthConfig)
    runtime: RuntimeConfig = field(default_factory=RuntimeConfig)
    profile: str = "full"
    capabilities: Capabilities = field(default_factory=Capabilities)
    custom: dict[str, str] = field(default_factory=dict)
    discover: DiscoverConfigSection = field(default_factory=DiscoverConfigSection)


_ENV_PATTERN = re.compile(r"\$\{([^}]+)\}")


def _interpolate(value: str) -> str:
    """Replace ${VAR} and ${VAR:-default} patterns with environment values."""

    def _replace(m: re.Match) -> str:
        expr = m.group(1)
        if ":-" in expr:
            var, default = expr.split(":-", 1)
            return os.environ.get(var, default)
        val = os.environ.get(expr)
        if val is None:
            raise ValueError(
                f"environment variable ${{{expr}}} is required but not set"
            )
        return val

    return _ENV_PATTERN.sub(_replace, value)


def _interpolate_recursive(obj: Any) -> Any:
    if isinstance(obj, str):
        return _interpolate(obj)
    if isinstance(obj, dict):
        return {k: _interpolate_recursive(v) for k, v in obj.items()}
    if isinstance(obj, list):
        return [_interpolate_recursive(v) for v in obj]
    return obj


def load_env(config_path: str | Path = "config.yaml") -> Env:
    """Load Env from config.yaml (with env interpolation) or pure env vars."""
    path = Path(config_path)

    if path.exists():
        raw = yaml.safe_load(path.read_text())
        if raw is None:
            raw = {}
        data = _interpolate_recursive(raw)
    else:
        data = _build_from_env()

    return _parse_env(data)


def _build_from_env() -> dict:
    """Fallback: construct config dict purely from environment variables."""
    return {
        "service": {
            "name": os.environ.get("E2E_SERVICE_NAME", ""),
            "base_url": os.environ.get("E2E_BASE_URL", ""),
        },
        "auth": {
            "tenant_id": os.environ.get("E2E_TENANT_ID", ""),
            "user_id": os.environ.get("E2E_USER_ID", ""),
        },
        "runtime": {},
        "profile": os.environ.get("E2E_PROFILE", "full"),
        "custom": {},
    }


def _parse_env(data: dict) -> Env:
    svc = data.get("service", {})
    auth_raw = data.get("auth", {})
    rt = data.get("runtime", {})
    custom = data.get("custom", {})
    disc = data.get("discover", {}) or {}

    auth_headers = auth_raw.get("headers") or {}
    auth_extra = {
        k: v
        for k, v in auth_raw.items()
        if k not in ("tenant_id", "user_id", "headers")
    }

    return Env(
        service=ServiceConfig(
            name=svc.get("name", ""),
            base_url=svc.get("base_url", "").rstrip("/"),
        ),
        auth=AuthConfig(
            tenant_id=auth_raw.get("tenant_id", ""),
            user_id=auth_raw.get("user_id", ""),
            headers={str(k): str(v) for k, v in auth_headers.items()}
            if auth_headers
            else {},
            extra=auth_extra,
        ),
        runtime=RuntimeConfig(
            http_timeout_s=int(rt.get("http_timeout_s", 120)),
            poll_interval_ms=int(rt.get("poll_interval_ms", 500)),
            poll_timeout_s=int(rt.get("poll_timeout_s", 60)),
            parallel=int(rt.get("parallel", 4)),
        ),
        profile=data.get("profile", "full"),
        custom={str(k): str(v) for k, v in custom.items()} if custom else {},
        discover=DiscoverConfigSection(
            source_root=str(disc.get("source_root", "")),
            test_root=str(disc.get("test_root", "")),
        ),
    )

File: go/e2e/core/env.go
// Package core provides environment configuration loading for e2e tests.
package core

import (
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"

	"gopkg.in/yaml.v3"
)

type ServiceConfig struct {
	Name    string `yaml:"name"`
	BaseURL string `yaml:"base_url"`
}

type AuthConfig struct {
	Headers map[string]string `yaml:"headers"`
}

type RuntimeConfig struct {
	HTTPTimeoutS   int `yaml:"http_timeout_s"`
	PollIntervalMS int `yaml:"poll_interval_ms"`
	PollTimeoutS   int `yaml:"poll_timeout_s"`
	Parallel       int `yaml:"parallel"`
}

type Capabilities struct {
	Known bool
	Data  map[string]any
}

type Env struct {
	Service      ServiceConfig
	Auth         AuthConfig
	Runtime      RuntimeConfig
	Profile      string
	Capabilities Capabilities
	Custom       map[string]string
}

var envPattern = regexp.MustCompile(`\$\{([^}]+)\}`)

func interpolate(s string) (string, error) {
	var lastErr error
	result := envPattern.ReplaceAllStringFunc(s, func(match string) string {
		expr := match[2 : len(match)-1] // strip ${ and }
		if idx := strings.Index(expr, ":-"); idx >= 0 {
			varName := expr[:idx]
			defaultVal := expr[idx+2:]
			if val := os.Getenv(varName); val != "" {
				return val
			}
			return defaultVal
		}
		val := os.Getenv(expr)
		if val == "" {
			lastErr = fmt.Errorf("environment variable ${%s} is required but not set", expr)
		}
		return val
	})
	return result, lastErr
}

func interpolateMap(raw map[string]any) (map[string]any, error) {
	out := make(map[string]any, len(raw))
	for k, v := range raw {
		resolved, err := interpolateValue(v)
		if err != nil {
			return nil, err
		}
		out[k] = resolved
	}
	return out, nil
}

func interpolateValue(v any) (any, error) {
	switch val := v.(type) {
	case string:
		return interpolate(val)
	case map[string]any:
		return interpolateMap(val)
	case []any:
		out := make([]any, len(val))
		for i, item := range val {
			resolved, err := interpolateValue(item)
			if err != nil {
				return nil, err
			}
			out[i] = resolved
		}
		return out, nil
	default:
		return v, nil
	}
}

// LoadEnv loads configuration from a YAML file with env var interpolation.
// Falls back to environment variables if the file doesn't exist.
func LoadEnv(configPath string) (*Env, error) {
	data, err := os.ReadFile(configPath)
	if err != nil {
		if os.IsNotExist(err) {
			return loadFromEnvVars()
		}
		return nil, fmt.Errorf("read config %s: %w", configPath, err)
	}

	var raw map[string]any
	if err := yaml.Unmarshal(data, &raw); err != nil {
		return nil, fmt.Errorf("parse config %s: %w", configPath, err)
	}

	resolved, err := interpolateMap(raw)
	if err != nil {
		return nil, err
	}

	return parseEnv(resolved), nil
}

func loadFromEnvVars() (*Env, error) {
	headers, err := authHeadersFromEnv()
	if err != nil {
		return nil, err
	}
	return &Env{
		Service: ServiceConfig{
			Name:    os.Getenv("E2E_SERVICE_NAME"),
			BaseURL: os.Getenv("E2E_BASE_URL"),
		},
		Auth: AuthConfig{Headers: headers},
		Runtime: RuntimeConfig{
			HTTPTimeoutS:   120,
			PollIntervalMS: 500,
			PollTimeoutS:   60,
			Parallel:       4,
		},
		Profile: envOrDefault("E2E_PROFILE", "full"),
		Custom:  map[string]string{},
	}, nil
}

func parseEnv(data map[string]any) *Env {
	env := &Env{
		Runtime: RuntimeConfig{
			HTTPTimeoutS:   120,
			PollIntervalMS: 500,
			PollTimeoutS:   60,
			Parallel:       4,
		},
		Profile: "full",
		Custom:  map[string]string{},
	}

	if svc, ok := data["service"].(map[string]any); ok {
		env.Service.Name = strVal(svc, "name")
		env.Service.BaseURL = strings.TrimRight(strVal(svc, "base_url"), "/")
	}
	if auth, ok := data["auth"].(map[string]any); ok {
		env.Auth.Headers = make(map[string]string)
		if headers, ok := auth["headers"].(map[string]any); ok {
			for key, value := range headers {
				env.Auth.Headers[key] = fmt.Sprint(value)
			}
		}
	}
	if rt, ok := data["runtime"].(map[string]any); ok {
		if v := intVal(rt, "http_timeout_s"); v > 0 {
			env.Runtime.HTTPTimeoutS = v
		}
		if v := intVal(rt, "poll_interval_ms"); v > 0 {
			env.Runtime.PollIntervalMS = v
		}
		if v := intVal(rt, "poll_timeout_s"); v > 0 {
			env.Runtime.PollTimeoutS = v
		}
		if v := intVal(rt, "parallel"); v > 0 {
			env.Runtime.Parallel = v
		}
	}
	if p := strVal(data, "profile"); p != "" {
		env.Profile = p
	}
	if custom, ok := data["custom"].(map[string]any); ok {
		for k, v := range custom {
			env.Custom[k] = fmt.Sprint(v)
		}
	}
	return env
}

func authHeadersFromEnv() (map[string]string, error) {
	raw := os.Getenv("E2E_AUTH_HEADERS")
	if raw == "" {
		return map[string]string{}, nil
	}
	headers := map[string]string{}
	if err := json.Unmarshal([]byte(raw), &headers); err != nil {
		return nil, fmt.Errorf("parse E2E_AUTH_HEADERS as JSON object: %w", err)
	}
	return headers, nil
}

func strVal(m map[string]any, key string) string {
	if v, ok := m[key]; ok {
		return fmt.Sprint(v)
	}
	return ""
}

func intVal(m map[string]any, key string) int {
	v, ok := m[key]
	if !ok {
		return 0
	}
	switch n := v.(type) {
	case int:
		return n
	case float64:
		return int(n)
	case string:
		i, _ := strconv.Atoi(n)
		return i
	}
	return 0
}

func envOrDefault(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}
